docs/architecture/adr-008-ics20-implementation.md

```diff
@@ -0,0 +1,189 @@
+    {
+        /// bind_port defines a wrapper function for the PortKeeper's bind_port function.
+        fn bind_port(&self, port_id: PortId) -> Result<(), ICS20Error>;
+        /// set_port sets the portID for the transfer module.
+        fn set_port(&self, port_id: PortId) -> ();
+        /// authenticate_capability wraps the CapabilityKeeper's authenticate_capability function
+        fn authenticate_capability(&self, cap: Capability, name: &str) -> bool;
+        /// claim_capability allows the transfer module to claim a capability that IBC module
+        /// passes to it
+        fn claim_capability(&self, cap: Capability, name: &str) -> Result<(), ICS20Error>;
+    }
+
+    pub trait ICS20Reader<AccountId: Into<String>>:
+    PortReader
+    + AccountReader<AccountId>
+    {
+        /// is_bound checks if the transfer module is already bound to the desired port.
+        fn is_bound(&self, port_id: PortId) -> bool;
+        /// get_transfer_account returns the ICS20 - transfers AccountId.
+        fn get_transfer_account(&self) -> AccountId;
+        /// get_port returns the portID for the transfer module.
+        fn get_port(&self) -> Result<PortId, Error>;
+    }
+
+    pub trait BankKeeper<AccountId: Into<String>> {
+        /// This function should enable sending ibc fungible tokens from one account to another
+        fn send_coins(&self, from: AccountId, to: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function to enable  minting ibc tokens in a module
+        fn mint_coins(&self, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable burning of minted tokens
+        fn burn_coins(&self, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable transfer of tokens from the ibc module to an account
+        fn send_coins_from_module_to_account(&self, module: AccountId, to: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+        /// This function should enable transfer of tokens from an account to the ibc module
+        fn send_coins_from_account_to_module(&self, from: AccountId, module: AccountId, amt: Coin) -> Result<(), ICS20Error>;
+    }
+
+    pub trait AccountReader<AccountId: Into<String>> {
+        /// This function should return the account of the ibc module
+        fn get_module_account(&self) -> AccountId;
+        /// Returns the escrow account id for a port and channel combination
+        fn get_escrow_account(port_id: PortId, channel_id: ChannelId) -> AccountId;
+    }
+
+    pub trait ICS20Context: ICS20Keeper + ICS20Reader;
+```
+## Handling ICS20 Packets
+ICS20 messages are still a subset of channel packets so they should be handled as such
+
+The following handlers are recommended to be implemented in the ics20_fungible_token_transfer application in the ibc-rs crate
+
+These handlers will be executed in the module callbacks of any thirdparty IBCmodule that is implementing an ICS20 application on-chain
```

Comment:
I think the design of these handlers will depend on the design of the ICS26 `Module` trait. We might need to impose additional restrictions due to trait object safety. https://github.com/informalsystems/ibc-rs/pull/1838/files#r809237272